Add troubleshooting guide for how to free JS storage (#14908)
kyma-project/kyma: docs/04-operation-guides/troubleshooting/eventing/evnt-04-free-jetstream-storage.md

---

> title: Free NATS JetStream file storage when it gets full

Author: I took you suggestion and appened `fue to full storage`, as it is the only reason we are talking about in this troubleshoot guide

---

## Symptom

> Free NATS JetStream file storage when it gets full.

Author: Yes, you are right, I changed the structure and applied your suggestions in the 07a0dd0

## Cause

> NATS JetStream uses [Interest retention policy](https://docs.nats.io/using-nats/developer/develop_jetstream/model_deep_dive) in Kyma per default.

Author: done

> It means, that as long as there are consumers on the stream, which the published event's subject, the messages will be kept in the stream if they cannot be delivered to the sink.

Author: done

> In some cases, it might happen, that the NATS JetStream storage gets full due to too many undelivered events.

Reviewer: @NHingerl Just changed a word. What do you think?

```suggestion
If there are too many undelivered events, the NATS JetStream storage may get full. To prevent event loss, the backend stops receiving events, and no further events can be persisted in the stream.
```

> In some cases, it might happen, that the NATS JetStream storage gets full due to too many undelivered events.

Author: I took the @NHingerl idea, would that be ok @raypinto ?

> In some cases, it might happen, that the NATS JetStream storage gets full due to too many undelivered events.

Reviewer: I think that messages are persisted **in** the stream and not **to** the stream. But @NHingerl could suggest better :)

> In some cases, it might happen, that the NATS JetStream storage gets full due to too many undelivered events.

Reviewer 2: That seems like a technical question rather than linguistic, I'll go with Ray's suggestion. I may have misunderstood where/how the persisting happens.

In order to not lose events, JetStream backend will just stop receiving and the user will get `507 Insufficient Storage` Error from Publisher Proxy or `no space left on device` from the Backend directly.
This means, that the Backend's storage is full and no further events can be persisted to the stream.

## Remedy

There are several ways of how to free the space on NATS JetStream Backend:

> - The published events might be too large, so the consumer isn't fast enough to deliver them, before the storage gets full. In that case eiter wait until the events get delivered or extend the NATS Backend by additional replicas.

Reviewer: @NHingerl changed one word here as well:

```suggestion
- If the published events are too large, the consumer cannot deliver them fast enough before the storage is full.
  In that case, either wait until the events are delivered, or scale the NATS backend with additional replicas.
```

> - The published events might be too large, so the consumer isn't fast enough to deliver them, before the storage gets full. In that case eiter wait until the events get delivered or extend the NATS Backend by additional replicas.

Author: @NHingerl, @raypinto I applied both of your suggestions, what do you think ?

> - Check if the sink is reachable and can accept the events.

Author: I like the idea, added one more link to the existing docs

> - Due to `Interest` Policy, the events published to the subject, which doesn't match any consumer filter, will not be kept in the stream.

Author: This is already a remedy option. So there in the first sentence I explain why the deletion of Subscription will help and then advice to delete it. Check the new version here 07a0dd0.

>   You can delete a Kyma Subscription, which will automatically remove all the pending messages in the stream, which were published the Subscription's subject. 

Reviewer: @NHingerl added a missing word (to). And yes it doesn't matter which Subscription is deleted. When you delete a subscription, it's associated messages in the stream (that were addressed to the eventTypes mentioned in the subscription) are removed. 

```suggestion
  To fix this, delete a Kyma Subscription, which automatically removes all the pending messages in the stream that were published to that Subscription's subject.
```

>   You can delete a Kyma Subscription, which will automatically remove all the pending messages in the stream, which were published the Subscription's subject. 

Author: @NHingerl, @raypinto thanks for suggestion, applied them both.

---
title: Eventing backend stopped receiving events due to full storage
---

## Symptom

NATS JetStream backend stopped receiving events due to full storage.

You see one of the following messages:
- `507 Insufficient Storage` error from Event Publisher Proxy.
- `no space left on device` from the eventing backend.

## Cause

In Kyma, the default retention policy for NATS JetStream is [Interest](https://docs.nats.io/using-nats/developer/develop_jetstream/model_deep_dive).
This retention policy keeps messages in the stream if they can't be delivered to the sink, as long as there are consumers in the stream that match the published event's subject.

If there are too many undelivered events, the NATS JetStream storage may get full.
To prevent event loss, the backend stops receiving events, and no further events can be persisted to the stream.

> >**NOTE:** Remember, that if you delete a Subscriber(sink), but there is still a Kyma Subscription pointing to that sink in place, the events published to that Subscription will pile up in the stream and possibly delay the event delivery to other Subscribers.

Reviewer 2: ```suggestion
> **NOTE:** If you delete a Subscriber (sink) while there is still a Kyma Subscription pointing to that sink, the events published to that Subscription pile up in the stream and possibly delay the event delivery to other Subscribers.
```
